src/message.rs

```diff
@@ -11,14 +12,46 @@ use fvm_shared::{
 use multihash_codetable::{Code, MultihashDigest as _};
 use serde::{Deserialize, Serialize};
 
-fn from_cbor_blake2b256<S: serde::ser::Serialize>(obj: &S) -> Result<Cid, Error> {
+fn from_cbor_blake2b256<S: serde::ser::Serialize>(
+    obj: &S,
+) -> Result<Cid, fvm_ipld_encoding::Error> {
     let bytes = fvm_ipld_encoding::to_vec(obj)?;
     Ok(Cid::new_v1(
         fvm_ipld_encoding::DAG_CBOR,
         Code::Blake2b256.digest(&bytes),
     ))
 }
 
+fn check_address_prefix(s: &str, n: Network) -> bool {
+    if s.len() == 0 || s.len() < 2 {
+        return false;
+    }
+
+    match n {
+        Network::Mainnet => s[0..1].eq("f") || s[0..2].eq("0x"),
+        Network::Testnet => s[0..1].eq("t") || s[0..2].eq("0x"),
+    }
+}
+
+pub fn parse_address(s: &str, n: Network) -> anyhow::Result<Address> {
+    if !check_address_prefix(s, n) {
+        return Err(anyhow!("Wrong Network"));
+    }
+
+    return match n.parse_address(s) {
+        Ok(addr) => Ok(addr),
+        Err(_e) => {
+            // Try parsing as 0x ethereum address
+            if s.len() != 42 {
+                return Err(anyhow!("Invalid Address"));
```

Comment:
nit: it would be nice to have more helpful error messages, given those are propagated to the user, e.g., invalid length, incorrect prefix etc. - we're already checking all of this anyway. Could be done in a follow-up, not a blocker.

Reply:
At this point in the code we don't actually know what the issue is. The length check informs us if it _could_ be an Ethereum address, but it also possible that something else failed in the above call to `n.parse_address()`. 

Open to suggestions, but IMO we can't win with error handling here. If we want good errors we should write our own validation logic  or improve `fvm_shared` 🙃

Reply:
What if we invert the logic? Some pseudo code

```
if prefix == '0x'
  parse_eth(addr)
else
  regular_parse(addr)
end
```

The regular parsing from FVM already gives us the proper error messages. Inside our custom `parse_eth` (you already have the code, just not wrapped in a function) you can do those checks and return proper errors, e.g.:
```
if addr.len != EXPECTED
  bail! "incorrect Ethereum address length - must be EXPECTED"
end

if addr.decode_hex.failed
  bail! invalid characters in Ethereum address
end
```

(more specifically, you might be interested in the `context()` method that `anyhow` exposes to make the code more concise)

What do you think? It's an improvement, though. We can do it in a follow-up or have a tracking issue. Up to you.

Reply:
You have a valid point that the logic should be inverted. Gone ahead and implemented that.

IMO best to merge this and let it be. Perfection is the enemy of progress, and contrary to CDD.